### PR TITLE
Tweak Doctrine dependencies and add a DBAL 3 compat fix

### PR DESCRIPTION
**Pull request by mbabker**
`doctrine/persistence` wasn't declared in the dependencies, added it with versions including the `Doctrine\Persistence` namespace

There isn't a big reason for `doctrine/orm` to be pinned to the last 2.9 patch release that I'm aware of (the XML linting should always be done on a setup with the latest supported version installed, so closing down the range just for that step is a little too strict).  I've also added a conflict to older ORM versions since effectively whatever is in dev is the minimum supported version for an optional dependency.

A conflict to `doctrine/cache:<1.11` is added to make sure that only versions of that library with the 2.0 compatibility layer are usable.

DBAL 3 compatibility is improved with this PR, and generally, projects supporting both DBAL 2 and 3 are locking to 2.13 thanks to the forward compatibility layer.

There's no need to conflict against `doctrine/mongodb` anymore.  That library is abandoned anyway and this library realistically doesn't support it anymore to begin with (with https://github.com/doctrine-extensions/DoctrineExtensions/pull/2260 removing the last remnants of references to it).

### CODE REVIEW

**Review by phansys — APPROVED** [review]

**Comment by phansys** [comment]
IMO, the only controversial change here is the conflict against "doctrine/dbal" >= 3.
Is there any way to avoid that?

**Comment by mbabker** [comment]
> Is there any way to avoid that?

Not without addressing the deprecations.  I quickly installed `symfony/phpunit-bridge` locally and ran the tests with that against this branch, here's the output (some of this is going to be false positives since it's vendor-calling-vendor deprecations, but at least it gives a decent snapshot with anything that has test coverage).  The PhpStorm inspections also pick up some uses of deprecated `fetch*` methods from the DBAL classes that aren't in test coverage, so as things sit right now, the package isn't DBAL 3 compatible.

```sh
There were 2 warnings:

1) Gedmo\Tree\MaterializedPathORMRepositoryTest::testIssue458
assertRegExp() is deprecated and will be removed in PHPUnit 10. Refactor your code to use assertMatchesRegularExpression() instead.

2) Gedmo\Uploadable\UploadableEntityTest::testFileWithFilenameSha1Generator
assertRegExp() is deprecated and will be removed in PHPUnit 10. Refactor your code to use assertMatchesRegularExpression() instead.

WARNINGS!
Tests: 455, Assertions: 2977, Warnings: 2, Skipped: 3.

Remaining indirect deprecation notices (17)

  5x: Since doctrine/mongodb-odm 2.1: The "boolean" mapping type is deprecated. Use "bool" instead.
    3x in TimestampableDocumentTest::setUp from Gedmo\Timestampable
    2x in IpTraceableDocumentTest::setUp from Gedmo\IpTraceable

  2x: Since doctrine/mongodb-odm 2.2: The "indexes" parameter in the "Gedmo\Loggable\Document\LogEntry" annotation for class "Doctrine\ODM\MongoDB\Mapping\Annotations\Document" is deprecated. Specify all "@Index" and "@UniqueIndex" annotations on the class.
    1x in LoggableDocumentTest::testLogGeneration from Gedmo\Loggable
    1x in LoggableDocumentTest::testVersionControl from Gedmo\Loggable

  1x: Method Doctrine\ORM\Configuration::getMetadataCacheImpl() is deprecated and will be removed in Doctrine ORM 3.0. Use getMetadataCache() instead. (Configuration.php:305 called by EntityManager.php:1018, https://github.com/doctrine/orm/issues/8650, package doctrine/orm)
    1x in EntityTranslationTableTest::setUp from Gedmo\Translatable

  1x: AbstractPlatform::prefersSequences() is deprecated without replacement and removed in DBAL 3.0 (AbstractPlatform.php:2676 called by ClassMetadataFactory.php:559, https://github.com/doctrine/dbal/pull/4229, package doctrine/dbal)
    1x in EntityTranslationTableTest::setUp from Gedmo\Translatable

  1x: Doctrine\Persistence\Mapping\AbstractClassMetadataFactory::getCacheDriver is deprecated. Use getCache() instead. (AbstractClassMetadataFactory.php:108 called by ExtensionMetadataFactory.php:104, https://github.com/doctrine/persistence/issues/184, package doctrine/persistence)
    1x in EntityTranslationTableTest::setUp from Gedmo\Translatable

  1x: Statement::execute() is deprecated, use Statement::executeQuery() or Statement::executeStatement() instead (Statement.php:166 called by BasicEntityPersister.php:287, https://github.com/doctrine/dbal/pull/4580, package doctrine/dbal)
    1x in EntityTranslationTableTest::testFixtureGeneratedTranslations from Gedmo\Translatable

  1x: Statement::closeCursor() is deprecated, use Result::free() instead. (Statement.php:243 called by BasicEntityPersister.php:305, https://github.com/doctrine/dbal/pull/4049, package doctrine/dbal)
    1x in EntityTranslationTableTest::testFixtureGeneratedTranslations from Gedmo\Translatable

  1x: Result::fetch() is deprecated, use Result::fetchNumeric(), fetchAssociative() or fetchOne() instead. (Result.php:89 called by ArrayHydrator.php:77, https://github.com/doctrine/dbal/pull/4019, package doctrine/dbal)
    1x in EntityTranslationTableTest::testFixtureGeneratedTranslations from Gedmo\Translatable

  1x: Connection::executeUpdate() is deprecated, use Connection::executeStatement() instead. (Connection.php:1483 called by BasicEntityPersister.php:512, https://github.com/doctrine/dbal/pull/4163, package doctrine/dbal)
    1x in Issue114Test::testIssue114 from Gedmo\Translatable

  1x: YAML mapping driver is deprecated and will be removed in Doctrine ORM 3.0, please migrate to annotation or XML driver. (YamlDriver.php:61 called by Issue116Test.php:38, https://github.com/doctrine/orm/issues/8465, package doctrine/orm)
    1x in Issue116Test::setUp from Gedmo\Sluggable

  1x: NOTIFY Change Tracking policy used in "Sortable\Fixture\NotifyNode" is deprecated, use deferred explicit instead. (ClassMetadataFactory.php:242 called by AbstractClassMetadataFactory.php:382, https://github.com/doctrine/orm/issues/8383, package doctrine/orm)
    1x in SortableTest::setUp from Gedmo\Sortable

  1x: Method Doctrine\ORM\EntityManager::merge() is deprecated and will be removed in Doctrine ORM 3.0. (EntityManager.php:723 called by TimestampableTest.php:49, https://github.com/doctrine/orm/issues/8461, package doctrine/orm)
    1x in TimestampableTest::shouldHandleDetatchedAndMergedBackEntities from Gedmo\Timestampable

Other deprecation notices (92)

  23x: The Doctrine\ORM\Configuration::getMetadataCacheImpl method is deprecated (Deprecated in favor of getMetadataCache).
    4x in MultiManagerMappingTest::setUp from Gedmo\Mapping
    4x in ReferencesListenerTest::setUp from Gedmo\References
    2x in LoggableMappingTest::setUp from Gedmo\Mapping\Xml
    2x in TranslatableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping
    1x in SortableMappingTest::setUp from Gedmo\Sluggable
    1x in UploadableMappingTest::setUp from Gedmo\Mapping
    1x in ClosureTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in MaterializedPathTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in NestedTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SortableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in TimestampableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in UploadableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in LoggableMappingTest::setUp from Gedmo\Mapping\Yaml

  23x: The Doctrine\ORM\Configuration::getProxyDir method is deprecated (2.7 We're switch to `ocramius/proxy-manager` and this method isn't applicable any longer).
    4x in MultiManagerMappingTest::setUp from Gedmo\Mapping
    4x in ReferencesListenerTest::setUp from Gedmo\References
    2x in LoggableMappingTest::setUp from Gedmo\Mapping\Xml
    2x in TranslatableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping
    1x in SortableMappingTest::setUp from Gedmo\Sluggable
    1x in UploadableMappingTest::setUp from Gedmo\Mapping
    1x in ClosureTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in MaterializedPathTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in NestedTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SortableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in TimestampableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in UploadableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in LoggableMappingTest::setUp from Gedmo\Mapping\Yaml

  23x: The Doctrine\ORM\Configuration::getProxyNamespace method is deprecated (2.7 We're switch to `ocramius/proxy-manager` and this method isn't applicable any longer).
    4x in MultiManagerMappingTest::setUp from Gedmo\Mapping
    4x in ReferencesListenerTest::setUp from Gedmo\References
    2x in LoggableMappingTest::setUp from Gedmo\Mapping\Xml
    2x in TranslatableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping
    1x in SortableMappingTest::setUp from Gedmo\Sluggable
    1x in UploadableMappingTest::setUp from Gedmo\Mapping
    1x in ClosureTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in MaterializedPathTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in NestedTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SortableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in TimestampableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in UploadableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in LoggableMappingTest::setUp from Gedmo\Mapping\Yaml

  23x: The Doctrine\ORM\Configuration::getAutoGenerateProxyClasses method is deprecated (2.7 We're switch to `ocramius/proxy-manager` and this method isn't applicable any longer).
    4x in MultiManagerMappingTest::setUp from Gedmo\Mapping
    4x in ReferencesListenerTest::setUp from Gedmo\References
    2x in LoggableMappingTest::setUp from Gedmo\Mapping\Xml
    2x in TranslatableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping
    1x in SortableMappingTest::setUp from Gedmo\Sluggable
    1x in UploadableMappingTest::setUp from Gedmo\Mapping
    1x in ClosureTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in MaterializedPathTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in NestedTreeMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SoftDeleteableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in SortableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in TimestampableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in UploadableMappingTest::setUp from Gedmo\Mapping\Xml
    1x in LoggableMappingTest::setUp from Gedmo\Mapping\Yaml
```

**Comment by franmomu** [comment]
> There isn't a big reason for `doctrine/orm` to be pinned to the last 2.9 patch release that I'm aware of (the XML linting should always be done on a setup with the latest supported version installed, so closing down the range just for that step is a little too strict). I've also added a conflict to older ORM versions since effectively whatever is in dev is the minimum supported version for an optional dependency.

I'm afraid that now it is because of https://github.com/doctrine-extensions/DoctrineExtensions/pull/2272, that change from `spl_object_hash` to `spl_object_hash` force us to add stricter conflict.

**Comment by mbabker** [comment]
Yeah, that one’s fair.  Being so tied to the internals, it’s hard not to make that change.

**Comment by mbabker** [comment]
I did one more scan after #2277 and the only other obvious break that came up was the change in PostgreSQL support.  I've added that to this PR, at face value there shouldn't be any DBAL 3 blockers left so the constraints will now allow it.

File: composer.json [hunk 1]
```diff
@@ -40,23 +40,24 @@
...
+        "doctrine/orm": "^2.9",
         "friendsofphp/php-cs-fixer": "^3.0",
         "phpunit/phpunit": "^8.5 || ^9.5",
         "symfony/cache": "^4.4 || ^5.0",
         "symfony/yaml": "^4.1"
     },
     "conflict": {
-        "doctrine/mongodb": "<1.3",
+        "doctrine/cache": "<1.11",
+        "doctrine/dbal": "<2.13.1 || ~3.0.0",
```
Comment (mbabker) [inline composer.json:58]: The conflicts here match up with some projects I maintain and what Symfony uses.  2.13.0 had some broken compat code, and IIRC 3.1.0 made some improvements on the compat layer so 3.0.x isn't really strongly recommended.

**Comment by mbabker** [comment]
Conflicts fixed.

**Comment by mbabker** [comment]
The CI failures will be fixed by https://github.com/doctrine-extensions/DoctrineExtensions/pull/2284

Since the branch for that was based on a combo of this PR, the ORM 2.10 PR, and my other changes, what I'll probably end up doing is rebasing the newer PR and closing this one in favor of the other one since it has some other DBAL 3 compat fixes for the tests and some general API cleanup.

**Comment by mbabker** [comment]
Closing in favor of #2284

**Pull request closed without merging**